packages/effect/src/ReadonlyRecord.ts

```diff
@@ -17,16 +17,37 @@ import type { NoInfer } from "./Types.js"
  * @category models
  * @since 2.0.0
  */
-export interface ReadonlyRecord<out A> {
-  readonly [x: string]: A
+export type ReadonlyRecord<in out K extends string | symbol, out A> = {
+  readonly [P in K]: A
+}
+
+/**
+ * @since 2.0.0
+ */
+export declare namespace ReadonlyRecord {
+  /**
+   * @since 2.0.0
+   */
+  export type NonLiteralKey<K extends string | symbol> = (
+    | (K extends string ? string : never)
+    | (K extends symbol ? symbol : never)
+  ) extends infer X ? X extends string | symbol ? X : never : never
+
+  /**
+   * @since 2.0.0
+   */
+  export type IntersectKeys<K1 extends string | symbol, K2 extends string | symbol> = [string] extends [K1 | K2] ?
+    NonLiteralKey<K1> & NonLiteralKey<K2>
+    : [symbol] extends [K1 | K2] ? NonLiteralKey<K1> & NonLiteralKey<K2>
+    : K1 & K2
 }
 
 /**
  * @category type lambdas
  * @since 2.0.0
  */
 export interface ReadonlyRecordTypeLambda extends TypeLambda {
-  readonly type: ReadonlyRecord<this["Target"]>
+  readonly type: ReadonlyRecord<string | symbol, this["Target"]>
```

Comment:
I'm afraid `string | symbol` will make the instances in `typeclass` less useful, for example

```ts
import * as ReadonlyRecordInstances from "@effect/typeclass/data/ReadonlyRecord"

const map = ReadonlyRecordInstances.Covariant.map
declare const r: Record<string, number>
// const result: ReadonlyRecord<string | symbol, number>
const result = map(r, (n) => n)
```

Maybe we could make `ReadonlyRecordTypeLambda` somehow parametric, I'll try to make some attempts